File: python/lira/arch_utils.py
from lira.arch import *

from typing import Dict

@dataclass
class ArchIndex:
    rf: Dict[str, RegisterFile]
    sr: Dict[str, SystemRegister]
    env: Dict[str, EnvironmentFunction]
    tables: Dict[str, TableInt]
    op: Dict[str, Operation]
    snippet: Dict[str, Snippet]
    instr: Dict[str, Instruction]

def build_arch_index(arch: Arch) -> ArchIndex:
    """Build index dictionaries from an Arch."""
    return ArchIndex(
        rf={rf.name: rf for rf in arch.register_files},
        sr={sr.name: sr for sr in arch.system_registers},
        env={ef.name: ef for ef in arch.environment_functions},
        tables={ti.name: ti for ti in arch.tables_int},
        op={op.name: op for op in arch.operations},
        snippet={sn.name: sn for sn in arch.snippets},
        instr={ins.name: ins for ins in arch.instructions},
    )

File: ruby/lira/arch_utils.rb
require_relative 'arch'

module LIRA
  ArchIndex = Struct.new(:rf, :sr, :env, :tables, :op, :snippet, :instr, keyword_init: true)

  def self.build_arch_index(arch)
    ArchIndex.new(
      rf: arch.register_files.to_h { |rf| [rf.name, rf] },
      sr: arch.system_registers.to_h { |sr| [sr.name, sr] },
      env: arch.environment_functions.to_h { |env| [env.name, env] },
      tables: arch.tables_int.to_h { |t| [t.name, t] },
      op: arch.operations.to_h { |op| [op.name, op] },
      snippet: arch.snippets.to_h { |sn| [sn.name, sn] },
      instr: arch.instructions.to_h { |ins| [ins.name, ins] }
    )
  end
end
